File: validate_reader.py
"""
validate_reader.py
------------------
Run this to confirm the reader and spread engine work correctly.

Usage:
    python validate_reader.py CL_outrights_1min_t.csv
    python validate_reader.py "C:\full\path\to\CL_outrights_1min_t.csv"
"""

import argparse
import sys
from pathlib import Path

import pandas as pd

sys.path.insert(0, str(Path(__file__).parent / "src"))

from demsup.data.futures_reader import FuturesFile, read_futures_csv
from demsup.spreads.time_spreads import SpreadEngine


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("path", help="Path to CSV file")
    parser.add_argument("--resample", default="1h",
                        help="Resample frequency (default: 1H)")
    args = parser.parse_args()

    path = Path(args.path)
    if not path.exists():
        print(f"ERROR: File not found: {path}")
        sys.exit(1)

    # ── Step 1: Raw reader ────────────────────────────────────────────────────
    print("\n" + "="*60)
    print("STEP 1: Raw reader")
    print("="*60)
    df_raw = read_futures_csv(path)
    print(f"\nShape      : {df_raw.shape}")
    print(f"Date range : {df_raw.index.min()} → {df_raw.index.max()}")
    print(f"Index tz   : {df_raw.index.tz}")
    print(f"Columns (first 6): {df_raw.columns.tolist()[:6]}")

    # ── Step 2: FuturesFile wrapper ───────────────────────────────────────────
    print("\n" + "="*60)
    print("STEP 2: FuturesFile wrapper")
    print("="*60)
    ff = FuturesFile(path)
    print(f"\n{ff}")

    # ── Step 3: Price extraction ──────────────────────────────────────────────
    print("\n" + "="*60)
    print(f"STEP 3: Price extraction (weighted_mid, resampled to {args.resample})")
    print("="*60)
    prices = ff.prices(resample_to=args.resample)
    print(f"\nShape    : {prices.shape}")
    print(f"Contracts: {prices.columns.tolist()}")
    print(f"\nLast 3 rows:")
    print(prices.tail(3).to_string())

    # ── Step 4: Spreads ───────────────────────────────────────────────────────
    print("\n" + "="*60)
    print("STEP 4: Time spreads")
    print("="*60)
    spds = ff.spreads(resample_to=args.resample)
    print(f"\nSpreads computed: {list(spds.columns)}")
    print(f"\nLast 5 rows:")
    print(spds.tail(5).to_string())

    # ── Step 5: Curve metrics ─────────────────────────────────────────────────
    print("\n" + "="*60)
    print("STEP 5: Curve metrics")
    print("="*60)
    ts = ff.term_structure(resample_to=args.resample)
    print(f"\nCurve columns: {list(ts['curve'].columns)}")
    print(f"\nLast 5 rows:")
    print(ts["curve"].tail(5).to_string())

    # ── Step 6: SpreadEngine z-scores ─────────────────────────────────────────
    print("\n" + "="*60)
    print("STEP 6: SpreadEngine — z-scores and signals")
    print("="*60)
    engine = SpreadEngine(ff, field="weighted_mid")
    bundle = engine.compute_all(resample_to=args.resample)

    current = bundle["current"]
    print(f"\n{'Spread':<12} {'Value':>8} {'Z-score':>8} {'Pct rank':>9}  Signal")
    print("-" * 65)
    for k in ["M1M2", "M1M3", "M1M6", "M1M12", "slope"]:
        if k not in current:
            continue
        c   = current[k]
        val = f"{c['value']:.3f}"        if c.get("value")    is not None else "N/A"
        z   = f"{c['z_score']:.2f}"      if c.get("z_score")  is not None else "N/A"
        pct = f"{c['pct_rank']:.0%}"     if c.get("pct_rank") is not None else "N/A"
        sig = c.get("signal", "N/A")
        print(f"{k:<12} {val:>8} {z:>8} {pct:>9}  {sig}")

    # ── Step 7: Data quality ──────────────────────────────────────────────────
    print("\n" + "="*60)
    print("STEP 7: Data quality")
    print("="*60)
    issues = []
    raw_prices = ff.prices()   # no resample — check raw nulls
    null_pct = raw_prices.isnull().mean()
    for col, pct in null_pct.items():
        if pct > 0.30:
            issues.append(f"  INFO  {col}: {pct:.0%} null (far contract, normal)")
        elif pct > 0.05:
            issues.append(f"  WARN  {col}: {pct:.0%} null (check data)")

    diffs = pd.Series(df_raw.index).diff().dropna()
    large_gaps = diffs[diffs > pd.Timedelta("4h")]
    if len(large_gaps) > 0:
        issues.append(f"  INFO  {len(large_gaps)} gaps > 4 hours (weekends/holidays — normal)")

    print(f"\n  Roll events (c1 ticker changes): {len(ff.roll_dates())}")
    if issues:
        for i in issues:
            print(i)
    else:
        print("  No data quality issues found.")

    print("\n" + "="*60)
    print("All steps complete. Reader and spread engine working correctly.")
    print("="*60)


if __name__ == "__main__":
    main()

File: validate_reader.R
# validate_reader.R
# ------------------
# Run this to confirm the R reader works against your actual CSV files.
#
# Usage (in Codespace terminal):
#   Rscript validate_reader.R /path/to/CL_outrights_1min_t.csv

library(data.table)
library(lubridate)
library(zoo)

source("R/futures_reader.R")

args <- commandArgs(trailingOnly = TRUE)
path <- if (length(args) > 0) args[1] else stop("Usage: Rscript validate_reader.R <path_to_csv>")

cat("\n", strrep("=", 60), "\n")
cat("STEP 1: Raw reader\n")
cat(strrep("=", 60), "\n")

ff <- read_futures_csv(path)
cat("\nShape      :", nrow(ff), "rows x", ncol(ff), "cols\n")
cat("Date range :", format(min(ff$timestamp)), "->", format(max(ff$timestamp)), "\n")
cat("First 3 col names:", paste(colnames(ff)[1:4], collapse=", "), "...\n")

cat("\n", strrep("=", 60), "\n")
cat("STEP 2: FuturesFile summary\n")
cat(strrep("=", 60), "\n\n")
print(ff)

cat("\n", strrep("=", 60), "\n")
cat("STEP 3: Price extraction (hourly)\n")
cat(strrep("=", 60), "\n")

prices <- get_prices(ff, resample_to = "1 hour")
cat("\nShape    :", nrow(prices), "rows x", ncol(prices), "cols\n")
cat("Contracts:", paste(setdiff(colnames(prices), "timestamp"), collapse=", "), "\n")
cat("\nLast 3 rows:\n")
print(tail(prices[, 1:7], 3))

cat("\n", strrep("=", 60), "\n")
cat("STEP 4: Time spreads\n")
cat(strrep("=", 60), "\n")

spds <- get_spreads(ff, resample_to = "1 hour")
cat("\nSpreads computed:", paste(setdiff(colnames(spds), "timestamp"), collapse=", "), "\n")
cat("\nLast 5 rows:\n")
print(tail(spds, 5))

cat("\n", strrep("=", 60), "\n")
cat("STEP 5: Curve metrics\n")
cat(strrep("=", 60), "\n")

curve <- get_curve_metrics(ff, resample_to = "1 hour")
cat("\nMetrics:", paste(setdiff(colnames(curve), "timestamp"), collapse=", "), "\n")
cat("\nLast 5 rows:\n")
print(tail(curve, 5))

cat("\n", strrep("=", 60), "\n")
cat("STEP 6: Spread z-scores (63-bar rolling window)\n")
cat(strrep("=", 60), "\n")

# Rolling z-score for M1M2
window <- 63
m1m2   <- spds$M1M2
roll_mean <- zoo::rollmean(m1m2, window, fill = NA, align = "right")
roll_sd   <- zoo::rollapply(m1m2, window, sd, fill = NA, align = "right")
z_score   <- (m1m2 - roll_mean) / roll_sd
pct_rank  <- rank(m1m2, na.last = "keep") / sum(!is.na(m1m2))

latest <- list(
  value    = round(tail(m1m2[!is.na(m1m2)], 1), 4),
  z_score  = round(tail(z_score[!is.na(z_score)], 1), 3),
  pct_rank = round(tail(pct_rank[!is.na(pct_rank)], 1), 3)
)

cat("\nM1M2 current snapshot:\n")
cat("  Value    :", latest$value, "\n")
cat("  Z-score  :", latest$z_score, "\n")
cat("  Pct rank :", scales::percent(latest$pct_rank, accuracy=1)
    |> tryCatch(error = function(e) paste0(round(latest$pct_rank*100), "%")), "\n")

# Curve regime signal
last_contango <- tail(curve$contango[!is.na(curve$contango)], 1)
last_slope    <- tail(curve$slope[!is.na(curve$slope)], 1)
regime_signal <- if (last_contango == 1) "CONTANGO" else "BACKWARDATION"

cat("\nCurve regime signal :", regime_signal, "\n")
cat("Current slope (M1-M6):", round(last_slope, 3), "\n")

cat("\n", strrep("=", 60), "\n")
cat("All steps complete. R reader working correctly.\n")
cat(strrep("=", 60), "\n\n")